R/derive_var_atoxgr.R

```diff
@@ -0,0 +1,287 @@
+
+  # Check metadata data set has required variables
+  assert_data_frame(
+    meta_criteria,
+    required_vars = vars(TERM, GRADE_CRITERIA_CODE, DIRECTION, SI_UNIT_CHECK, VAR_CHECK)
+  )
+
+  # check input parameter has correct value
+  assert_character_scalar(criteria_direction, values = c("L", "H"))
+
+
+  # Get list of terms from criteria metadata with particular direction
+  # L = low (Hypo) H = high (Hyper)
+  atoxgr_dir <- meta_criteria %>%
+    filter(!is.na(GRADE_CRITERIA_CODE) & toupper(DIRECTION) == toupper(criteria_direction)) %>%
+    select(TERM, DIRECTION, SI_UNIT_CHECK, GRADE_CRITERIA_CODE, VAR_CHECK) %>%
+    mutate(
+      TERM_UPPER = toupper(TERM),
+      SI_UNIT_UPPER = toupper(SI_UNIT_CHECK)
+    )
+
+  # from ADLB VAD get distinct list of terms to be graded
+  terms_in_vad <- dataset %>%
+    filter(!is.na(!!tox_description_var)) %>%
+    distinct(!!tox_description_var) %>%
+    mutate(
+      TERM = !!tox_description_var,
+      TERM_UPPER = toupper(TERM)
+    )
+
+  # only keep terms that exist in both ADLB data and criteria metadata
+  list_of_terms <- terms_in_vad %>%
+    semi_join(atoxgr_dir, by = "TERM_UPPER") %>%
+    arrange(TERM)
+
+  # output lab data not to be graded
+  # this will be appended to in for loop after each term is graded
+  out_data <- dataset %>%
+    filter(!!tox_description_var %notin% (list_of_terms$TERM) | is.na(!!tox_description_var)) %>%
+    mutate(!!new_var := NA_character_)
+
+  # get lab data to be graded
+  to_be_graded <- dataset %>%
+    filter(!!tox_description_var %in% (list_of_terms$TERM))
+
+  # for each TERM apply criteria and create grade derivation
+  for (i in seq_along(list_of_terms$TERM)) {
+
+    # filter metadata on a term
+    meta_this_term <- atoxgr_dir %>%
+      filter(TERM_UPPER == list_of_terms$TERM_UPPER[i])
+
+    # Put list of variables required for criteria in a vector
+    list_of_vars <- gsub("\\s+", "", unlist(strsplit(meta_this_term$VAR_CHECK, ",")))
+
+    # filter lab data on term and apply criteria to derive grade
+    grade_this_term <- to_be_graded %>%
+      filter(!!tox_description_var == list_of_terms$TERM[i])
+
+    # check variables required in criteria exist on data
+    assert_data_frame(grade_this_term, required_vars = vars(!!!syms(list_of_vars)))
+
+    # apply criteria when SI unit matches
+    grade_this_term <- grade_this_term %>%
+      mutate(
+        temp_flag = meta_this_term$SI_UNIT_UPPER == toupper(!!get_unit_expr) |
+          is.na(meta_this_term$SI_UNIT_UPPER),
+        !!new_var := if_else(
+          temp_flag, eval(parse(text = meta_this_term$GRADE_CRITERIA_CODE)), NA_character_
+        )
+      ) %>%
+      select(-temp_flag)
+
+    # remove lab data just graded from data still to be graded
+    to_be_graded <- to_be_graded %>%
+      filter(!!tox_description_var != list_of_terms$TERM[i])
+
+    # append lab data just graded to output data
+    out_data <- bind_rows(out_data, grade_this_term)
+  }
+
+  all_data <- out_data
```

Comment:
I think we should directly return the final data instead of assigning it to another object.

Reply:
Ben also mentioned this, its done now